boa/contracts/abi/contract.py

```diff
@@ -0,0 +1,120 @@
+from functools import cached_property
+from operator import attrgetter
+from os.path import basename
+from typing import Any, Optional, Union
+from warnings import warn
+
+from eth.abc import ComputationAPI
+
+from boa.contracts.abi import _decode_addresses, _format_abi_type, group_by
+from boa.contracts.abi.function import ABIFunction, ABIOverload
+from boa.contracts.evm_contract import BaseEVMContract
+from boa.environment import Address
+from boa.util.abi import abi_decode
+from boa.util.exceptions import StackTrace, _handle_child_trace
+
+
+class ABIContract(BaseEVMContract):
+    """A contract that has been deployed to the blockchain and created via an ABI."""
+
+    def __init__(
+        self,
+        name: str,
+        functions: list[ABIFunction],
+        address: Address,
+        filename: Optional[str] = None,
+        env=None,
+    ):
+        super().__init__(env, filename=filename, address=address)
+        self._name = name
+        self._functions = functions
+        self._bytecode = self.env.vm.state.get_code(address.canonical_address)
+        if not self._bytecode:
+            warn(
+                f"Requested {self} but there is no bytecode at that address!",
+                stacklevel=2,
+            )
+
+        for name, group in group_by(self._functions, attrgetter("name")).items():
```

Comment:
honestly the groupby function could be inlined here - i don't mind reading a loop which constructs the dictionary :)